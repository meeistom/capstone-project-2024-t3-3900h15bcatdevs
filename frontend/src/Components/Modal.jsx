import React, { useEffect, useRef, useState } from 'react';
import axios from 'axios'; 
import '../index.css';
import scanner from '../Assets/scanner.png';
import sticker from '../Assets/milk1_label.png'
import { Form } from './Form.jsx';
import confirmCheck from '../Assets/confirm-check.png';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faXmark } from '@fortawesome/free-solid-svg-icons';
import Button from 'react-bootstrap/Button';

export { Modal };

function Modal({ closeModal, version }) {
  const scannerInputRef = useRef(null);
  const [modalVersion, setModalVersion] = useState(version);
  const [scannedValue, setScannedValue] = useState(0);
  const [expiryDate, setExpiryDate] = useState('');
  const [expressDate, setExpressDate] = useState('')
  const [milkType, setMilkType] = useState('ehm');
  const [storageType, setStorageType] = useState('fridge');
  const [babyData, setBabyData] = useState(null);
  const [motherData, setMotherData] = useState(null);
  let title, body, footer;

  useEffect(() => {
    if (scannerInputRef.current) {
      scannerInputRef.current.focus();
      // console.log("Input field focused"); // for debug
    }
  }, []);

  const getBabyDetails = async (barcode) => {
    const url = `http://localhost:5001/babies?mrn=${barcode}`;
    console.log(url);
    try {
      const response = await axios.get(url);
      setBabyData(response.data);
      console.log('Baby data fetched:', response.data);
      return response.data; 
    } catch (error) {
      console.log('Error fetching baby data:', error);
      alert('Failed to fetch baby details with corresponding barcode. Please try again.');
      return null; 
    }
  };
  
  const getMotherDetails = async (mother_mrn) => {
    const url = `http://localhost:5001/mothers?mrn=${mother_mrn}`;
    console.log(url);
    try {
      const response = await axios.get(url);
      setMotherData(response.data);
      console.log('Mother data fetched:', response.data);
      setModalVersion('addMilk2'); 
    } catch (error) {
      console.log('Error fetching Mother data:', error);
      alert('Failed to fetch mother details with corresponding mrn. Please try again.');
    }
  };
  
  const handleInput = async (event) => {
    const barcode = event.target.value;
    console.log(barcode);
    setScannedValue(barcode);
    if (barcode.length === 4) {
      const babyDetails = await getBabyDetails(barcode);
      if (babyDetails && babyDetails.mother_mrn) {
        await getMotherDetails(babyDetails.mother_mrn);
      }
    }
  };

  const handleSubmitMilkInfo = () => {
    const bottleDetails = {
      milk_type: milkType,
      express_time: expressDate,
      expiration_time: expiryDate,
      storage_type: storageType,
      storage_location: 'level 1',
      volume_ml: '50',
      owner_mrn: babyData.mrn,
      extra_notes: ' ',
    };

    const url = 'http://localhost:5001/add_milk_entry';
    axios.post(url, bottleDetails)
      .then(response => {
        console.log('Bottle details added:');
        setModalVersion('addMilk3');
      })
      .catch(error => {
        console.log('Error posting bottle details:', error);
        alert('Failed to add bottle details. Please try again.');
      });
    
  }

  const printImage = () => {
    const printWindow = window.open('', '_blank');
    printWindow.document.write(`<img src="${sticker}" alt="sticker" style="max-width: 100%;" />`);
    printWindow.document.close();
    printWindow.print();
  };

  const handleCheckInput = () => {
    if (!expiryDate || !expressDate) {
      // for debug
      // console.log(expiryDate)
      // console.log(expressDate)
      alert('Please fill in all relevant infomation');
    } else {
      handleSubmitMilkInfo()
    }
  }

  const handlePrintAndMovePage = () => {
    printImage();
    setModalVersion('addMilk4');
  }

  

  // handle which version of modal is rendered
  switch (modalVersion) {
    case 'addMilk1':
      title = 'Please scan the baby\'s barcode';
      body = (
        <>
          <input
            type="text"
            className="scanner-input"
            ref={scannerInputRef}
            onChange={handleInput}
          ></input>
          <img src={scanner} alt="scanner" />
        </>
      );
      footer = 'Waiting for scan...';
      break;
    case 'addMilk2':
      title = 'Confirm Infomation';
      body = (
        <>
          <Form 
          babyData={babyData} 
          motherData={motherData} 
          setExpressDate={setExpressDate} 
          setExpiryDate={setExpiryDate}
          setMilkType={setMilkType} 
          setStorageType={setStorageType} 
          />
        </>
      )
      footer = (
        <div id="btn-group">
         <button onClick={() => closeModal(false)} type="button" class="btn btn-outline-primary">Cancel</button>
         <button type="button" class="btn btn-primary" onClick={handleCheckInput}>Preview Sticker</button>
        </div>
      )
      break;
    case 'addMilk3':
      title = 'Sticker Preview'
      body = <img src={sticker} alt="sticker" />
      footer = (
        <div id="btn-group">
          <button onClick={() => setModalVersion('addMilk2')} type="button" class="btn btn-outline-primary">Back to Edit</button>
          <button type="button" class="btn btn-primary" onClick={handlePrintAndMovePage}>Confirm and Print</button>
        </div>
      )
      break;
    case 'addMilk4':
      title = 'Milk Added Successfully';
      body = (
        <div className="milk-confirmed">
          <div>
            Milk from the Mother was added. You may close the pop up now.
          </div>
          <img
            src={confirmCheck}
            alt="Confirmation Icon"
            className="confirm-img"
          />
        </div>
      );
      footer = (
        <div id="btn-group">
          <button onClick={printImage} type="button" class="btn btn-outline-primary">Reprint</button>
          <Button onClick={() => closeModal(false)}>Return Home</Button>
        </div>
      )
      break;
  }

  return (
    <>
      <div className="modal-background">
        <div className="modal-container">
          <div className="modal-close-btn">
            <button onClick={() => closeModal(false)}>
              <FontAwesomeIcon className="close-btn" icon={faXmark} />
            </button>
          </div>
          <div className="title">
            <p>{title}</p>
          </div>
          <div className="body">{body}</div>
          <div className="footer">{footer}</div>
        </div>
      </div>
    </>
  );
}
