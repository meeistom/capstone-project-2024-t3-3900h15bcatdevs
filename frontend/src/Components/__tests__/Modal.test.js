// modal is not seen when closed
// modal is visible when open

import { render, screen, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import { Modal } from "../Modal";
import "@testing-library/jest-dom";

afterEach(() => {
  cleanup();
});

test("should render modal", () => {
  render(
    <Modal
      id="test"
      title="Test Modal"
      body={<p>Body</p>}
      footer={<p>Footer</p>}
      closeModal={() => {}}
    />
  );

  expect(screen.getByText("Test Modal")).toBeInTheDocument();
  expect(screen.getByText("Body")).toBeInTheDocument();
  expect(screen.getByText("Footer")).toBeInTheDocument();
});

test("should match snapshot", () => {
  const tree = renderer
    .create(
      <Modal
        id="test"
        title="Test Modal"
        body={<p>Body</p>}
        footer={<p>Footer</p>}
        closeModal={() => {}}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("should be able to close modal", async () => {
  const closeModalMock = jest.fn();

  render(
    <Modal
      id="test-modal"
      title="Test Title"
      body={<p>Body</p>}
      footer={<p>Footer</p>}
      closeModal={closeModalMock}
    />
  );

  const closeButton = screen.getByRole("button", { name: "close-modal" });

  await userEvent.click(closeButton);

  expect(closeModalMock).toHaveBeenCalledWith(false);
  expect(closeModalMock).toHaveBeenCalledTimes(1);
});

test("should not show title, body or footer if props not given", () => {
  render(<Modal id="test-modal" closeModal={() => {}} />);

  expect(screen.queryByTestId("title")).not.toBeInTheDocument();
  expect(screen.queryByTestId("body")).not.toBeInTheDocument();
  expect(screen.queryByTestId("footer")).not.toBeInTheDocument();
});
