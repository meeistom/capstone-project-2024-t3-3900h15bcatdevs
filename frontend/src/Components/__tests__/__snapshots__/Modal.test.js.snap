// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="modal-background position-fixed d-flex justify-content-center align-items-center w-100 h-100"
  id="test"
>
  <div
    className="modal-container position-relative bg-white rounded-3 shadow-lg d-flex flex-column p-4"
  >
    <div
      className="modal-close-btn d-flex justify-content-end"
    >
      <button
        aria-label="close-modal"
        className="btn btn-link"
        disabled={false}
        name="close-modal"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-xmark close-btn"
          data-icon="xmark"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
    <div
      className="title fs-2 text-center"
    >
      <p>
        Test Modal
      </p>
    </div>
    <div
      className="body d-flex justify-content-center align-items-center flex-column"
    >
      <p>
        Body
      </p>
    </div>
    <div
      className="footer d-flex justify-content-center mt-auto w-100 align-items-center"
    >
      <p>
        Footer
      </p>
    </div>
  </div>
</div>
`;
